exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,375 @@
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// Always attempt to Patch the ClusterResourceSetBinding object after each reconciliation.
+		if err := patchHelper.Patch(ctx, clusterResourceSetBinding); err != nil {
+			r.Log.Error(err, "failed to patch config")
+		}
+	}()
+
+	// Initializes the maps in ClusterResourceSetBinding object if they are uninitialized.
+	InitClusterResourceSetBinding(clusterResourceSetBinding, clusterResourceSet)
+
+	var clusterResourceSetInMap clusterresourcesetv1.ResourcesSetBinding
+	var ok bool
+	if clusterResourceSetInMap, ok = clusterResourceSetBinding.ClusterResourceSetMap[clusterResourceSet.Name]; !ok {
+		// This should never be the case as InitClusterResourceSetBinding is adding the ClusterResourceSet to the map during initialization.
+		return errors.New("failed to find ClusterResourceSet " + clusterResourceSet.Name + " in ClusterResourceBinding")
+	}
+
+	errList := []error{}
+
+	// Iterate all resources and apply them to the cluster and update the resource status in the ClusterResourceSetBinding object.
+	for _, resource := range clusterResourceSet.Spec.Resources {
+		// If resource is already applied successfully and clusterResourceSet mod is "ApplyOnce", continue. (No need to check hash changes here)
+		resourceInMap := clusterresourcesetv1.ResourceBinding{}
+		if resourceInMap, ok = clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)]; ok {
+			if clusterResourceSet.Spec.Mode == string(clusterresourcesetv1.ClusterResourceSetModeApplyOnce) && resourceInMap.Applied {
+				continue
+			}
+		}
+
+		// This is set here to cover unsuccessful apply scenarios (if the end of the loop is not reached)
+		resourceInMap = *GenerateResourceStatus(false, "")
+		clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)] = resourceInMap
+
+		// Only allow using resources (Secrets/Configmaps) in the same namespace with the cluster.
+		typedName := types.NamespacedName{Name: resource.Name, Namespace: cluster.Namespace}
+		dataList := make([][]byte, 0)
+
+		// Retrieve data in the resource as an array because there can be many key-value pairs in resource and all will be applied to the cluster.
+		switch resource.Kind {
+		case string(clusterresourcesetv1.ConfigMapClusterResourceSetResourceKind):
+			// Set status in ClusterResourceSetBinding in case of early continue due to a failure. Do this for only supported kinds.
+			clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)] = clusterresourcesetv1.ResourceBinding{
+				Hash:            "",
+				Applied:         false,
+				LastAppliedTime: &metav1.Time{Time: time.Now().UTC()},
+			}
+
+			resourceConfigMap, err := GetConfigMapFromNamespacedName(context.Background(), r.Client, typedName)
+			if err != nil {
+				logger.Error(err, "Failed to find ConfigMap resource", "ConfigMap name", typedName)
+				errList = append(errList, err)
+				continue
+			}
+
+			for _, dataStr := range resourceConfigMap.Data {
```

Comment:
In [20200220-cluster-resource-set.md](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200220-cluster-resource-set.md#story-3), user story 3, it mentioned it's helpful to provide template ability. Is the template mechanism implemented in this patch? It seems it loads the verbatim ConfigMap and Secret data.

Reply:
Looks like it might have been a leftover from an earlier version of the proposal, for now we're keeping it very simple, but I can see that as a possible improvement later on

Reply:
We are not planning to implement any sort of templating. If you need templating, I would advise that it be done "up front" - for example:

1. Create files containing your Cluster, KubeadmControlPlane, MachineDeployments, etc.
1. Create files containing the resources you want to apply with CRS. These can contain placeholders.
1. Set your placeholder values (e.g. as environment variables, or in a file that your template processor uses)
1. Run the files through your template processor (could be as simple as `envsubst`)
1. Apply the processed output to your apiserver

Reply:
Make sense. Thanks, vincepri, ncdc.